AppCenter/AppCenterTests/MSACUtilityTests.m

```diff
@@ -156,6 +156,42 @@ - (void)testNowInMilliseconds {
   XCTAssertGreaterThan(actual, 0);
 }
 
+- (void)testDateFormatterConcurrentInitialization {
+
+  // If
+  [MSACUtility resetDateFormatterInstance];
+  XCTestExpectation *expectation = [self expectationWithDescription:@"queueExpectation"];
+  dispatch_queue_t concurrentQueue = dispatch_queue_create("com.dateformatter.queue", DISPATCH_QUEUE_CONCURRENT);
+  id nsDateFormatter = OCMClassMock([NSDateFormatter class]);
+  OCMStub([nsDateFormatter stringFromDate:OCMOCK_ANY]).andReturn(@"stub");
+  OCMStub([nsDateFormatter alloc]).andReturn(nsDateFormatter);
+
+  // When
+  int dispatchTimes = 10;
+  for (int i = 0; i <= dispatchTimes; i++) {
+    dispatch_async(concurrentQueue, ^{
+      if (i == dispatchTimes) {
+        [expectation fulfill];
```

Comment:
Wouldn't we get here before even the first test starts? Code inside the `dispatch_async`'s will be called after we iterate 10 times and `i` becomes `10`, because it's async in another thread. And `[expectation fulfill]` will be called 10 times, because in each dispatch block `i == dispatchTimes` is `true`.

I think we should create a variable, that we would increase in each dispatch block and when that variable is equal to `dispatchTimes` we're calling fulfill (but after the `try` block, not before).

Please correct me if I'm wrong.

Reply:
The block is catching the i value when created, so each dispatch_async will have the corresponding value, e.g. 0,1,2,3, and so on. But since they are concurrent they are not guaranteed to run in any particular order. So, yes, it makes sense to increase the counter inside the dispatch block, I'll fix it.